### ISSUE TITLE
PHP | Performance | Overhead optimizations | Spike 2

### ISSUE THREAD

**Issue by adri** [newrelic/newrelic-php-agent#859]
## Description

When updating from 10.17.0.7 to 10.19.0.9 we saw an average response increase of ~10ms, from ~37ms to ~47ms. After reverting the response time decreased.

<img width="1013" alt="SCR-20240319-nqxl" src="https://github.com/newrelic/newrelic-php-agent/assets/133832/31f2904c-066f-4c80-8a05-2acb9d1f4505">

## Steps to Reproduce

I don't know

## Expected Behavior

No performance impact

## Your Environment

Symfony v7.0
PHP 8.3.2
Debian 12

**Comment by ruudk** [comment]
@newrelic-php-agent-bot What does that performance degradation has to do with our support plan. We are paying customers. We report that your latest SDK degrades performance. And the previous did not. Clearly something is off.

**Comment by zsistla** [comment]
@ruudk thanks for the question, and we'll try to clear up the intent. The performance degradation has nothing to do with the support plan and is an issue we are investigating.  Sometimes (as in this case), in order to investigate further, we need certain details from customers that cannot be shared in a public repo.  Our method for collecting this data in a manner to provide privacy and security to the customer is via our support channel which can be reached in various ways (as detailed in the bot message).

**Comment by slt** [comment]
We noticed this when our systems were upgraded from v10.18.0.7 to v10.18.0.8, our endpoints were performing dramatically worse, even basic nothing endpoints like health checks.

We build our base docker containers with the latest versions nightly so within a few days of the v10.18.0.8 release it was very clear something had gone wrong.

This is a health check endpoint that does nothing but return a string

![image](https://github.com/newrelic/newrelic-php-agent/assets/2295572/98e00e57-5476-4c47-8add-ffc1e43c1d23)

This is a more involved endpoint which is usually consistently <60ms

![image](https://github.com/newrelic/newrelic-php-agent/assets/2295572/7f3f4210-d749-49b3-bb3c-7955b4a1cf49)


Given the change log, we thought perhaps it was due to JIT finally being available. 

>
> New features
> 
>     PHP's Just-In-Time Compilation is no longer disabled when the PHP Agent is enabled. PHP has this on-by-default starting with PHP 8.0.
> 

We have since tested with many different configurations of JIT, including JIT off and it seems to us at least this performance regression is due v10.18.0.8 itself and not JIT.

We've rolled back and pinned to v10.18.0.7 and are watching this issue & release announcements hoping for a fix, just commenting here to give it weight.

**Comment by ebeltramo96** [comment]
We noticed the same as soon as we merged the new version, we had to roll it back to `v10.17.0.7`

**Comment by ACenterA** [comment]
10.20.0.10 has been released:

[10.20.0.10](https://github.com/newrelic/newrelic-php-agent/releases/tag/v10.20.0.10) 

Anyone knows if this bug still occurs with that version?

**Comment by ppp0** [comment]
Clearly to be seen after upgrading from 10.17.0.7 to 10.22.0.12, albeit not so dramatic as reported above (~5%)
![image](https://github.com/newrelic/newrelic-php-agent/assets/1888714/fb9e78f5-5761-4940-a567-a284c6b03ef6)

**Comment by ruudk** [comment]
> Would you be please be able to provide more details about your system: OS PHP version(s) frameworks in use
> 
> If you follow the [bot instructions](https://github.com/newrelic/newrelic-php-agent/issues/859#issuecomment-2010719932), our support team could work with you to collect additional information such as debug logs and other sensitive system info in a secure manner.


OS: Debian 12 bookworm
PHP: PHP 8.3.11
Framework: Symfony 7.1

@zsistla I just created a Case #00224017

**Comment by flokain** [comment]
thx for looking into this. This is blocking us from using distributed tracing, so we hope this is fixed soon.

**Comment by ebeltramo96** [comment]
is there any news on this tipic?

**Comment by Mat-Moo** [comment]
Are you seeing the same in 11 agent? (I suspect so)

**Comment by flokain** [comment]
yes it is we are a paying customer and have filed this with our account manager anna tauriello. and she said she will escalate this internally so please get in touch with her, for more details and logs about our tests.
what i can tell we used newrelic php agent  version `11.3.0.16z and then activated `newrelic.distributed_tracing_enabled = true` an `newrelic.infinite_tracing.trace_observer.host="******.aws-eu-central-1.tracing.edge.eu.nr-data.net`. The effect you can see in the below graph `average(duration) from transaction` increasing from 4ms 30ms
![image](https://github.com/user-attachments/assets/0ce27a39-aa5c-463f-849d-475d2def5a81)

**Comment by slt** [comment]
Any details on what this was / how it was resolved?  I assume this will be in a future upcoming release?   We are eager to update to a current version thanks.

**Comment by eng1neer** [comment]
I'd like to share my performance measurements that raise serious questions about the meaningfulness of the JIT support in newer agent versions:

Regular application code (real world use case – template generation):
- No NewRelic: ~1.34s
- Old NewRelic (10.17.0.7): ~1.76s (31% overhead)
- New NewRelic, JIT disabled: ~2.08s (**55% overhead**)
- New NewRelic, JIT enabled: ~2.0s (**49% overhead**)

For CPU-intensive workloads. Synthetic test that demonstrates JIT speedup (fibonacci calculation):
- No NewRelic, JIT OFF: ~4.5s
- No NewRelic, JIT ON: ~2.0s (demonstrates JIT's intended 2.25x speedup)
- Old NewRelic (10.17.0.7): ~14.0s (3.1x slower than baseline)
- New NewRelic, JIT disabled: ~34.6s (**7.7x slower than baseline**)
- New NewRelic, JIT enabled: ~31.7s (**7x slower than baseline, 15.9x slower than JIT-optimized code**)

This raises a fundamental question: What was the goal of implementing JIT support in NR PHP agent when it has made performance worse across the board? JIT is not a feature that needs "support" - it's a performance optimization that only makes sense as long as it makes things faster. If it does not makes things faster, that support should not be there in the first place. The new observer API implementation not only makes regular code slower than the old version (55% vs 31% overhead), but makes CPU-intensive JIT-optimizable code dramatically worse regardless of JIT setting.

For now, we're staying on version 10.17.0.7. Could you provide configuration options to disable the observer API and fall back to the previous implementation? If the overhead cannot be significantly reduced, we would be better off with the older pre-jit implementation.

**Comment by ebeltramo96** [comment]
hello, @lavarou is there a follow up on your statement?

**Comment by lavarou** [comment]
> hello, [@lavarou](https://github.com/lavarou) is there a follow up on your statement?

Hello @ebeltramo96, @slt, @ruudk and @adri! New Relic has no control over implementation and performance of Zend Engine’s hooks made available to APM extensions like New Relic PHP Agent. Zend Engine core team themselves is aware of the overhead caused by those hooks and are also working on optimizations - PHP 8.4 release included [this change](https://github.com/php/php-src/pull/13649). However, as [mentioned earlier](https://github.com/newrelic/newrelic-php-agent/issues/859#issuecomment-2400564716), New Relic PHP Engineering team looked into options to reduce the overhead when [observer API](https://github.com/php/php-src/blob/PHP-8.0/Zend/zend_observer.h) hooks are used, and [PHP agent v11.9.0.23](https://docs.newrelic.com/docs/release-notes/agent-release-notes/php-release-notes/php-agent-11-9-0-23/) release includes a few optimizations of how New Relic PHP Agent uses those hooks. Please give it a try.

**Comment by ebeltramo96** [comment]
thanks @lavarou for the reply, we updated the php agent to the latest version but unfortunately we didn't see any improvements in the dashboards, we are running php version 8.2.
